### PR TITLE
CORTX-29207: Script to run code coverage via Jenkins

### PR DESCRIPTION
**Pull request by rkothiya**
This script is used in a jenkins job to do the code
coverage of the motr code.

Signed-off-by: Rinku Kothiya <rinku.kothiya@seagate.com>

# Problem Statement
Create a new job to run code coverage for cortx-motr

# Design
This script will run the code coverage on motr code.

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: scripts/jenkins/code-coverage.groovy [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; ./configure --enable-coverage --enable-debug"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; make -j6"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; sudo ./scripts/m0 run-ut"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; ./scripts/coverage/gcov-gen-html user ./ ./"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cp -r /root/cortx-motr/* /var/www/html/"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "firewall-cmd --permanent --add-port=80/tcp"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "firewall-cmd --permanent --add-port=443/tcp"
```
Comment (rkothiya) [inline scripts/jenkins/code-coverage.groovy:45]: If the firewall is not running then this step will fail, which needs to be addressed

Comment (rkothiya) [reply scripts/jenkins/code-coverage.groovy:49]: Now that we are starting the firewall manually this should not be a problem.

File: scripts/jenkins/code-coverage.groovy [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                    echo "Old build number is : $OLD_BUILD_NUMBER"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "rm -rf /var/www/html/$OLD_BUILD_NUMBER || true"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/ ; git clone --recursive -b ${BRANCH} https://github.com/Seagate/cortx-motr"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; sudo ./scripts/install-build-deps"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; ./autogen.sh"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; ./configure --enable-coverage --enable-debug"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; make -j6"
+                    sshpass -p ${NODE_PASS} ssh -o StrictHostKeyChecking=no ${NODE_USER}@${NODE_HOST} "cd /root/cortx-motr ; sudo ./scripts/m0 run-ut"
```
Comment (rkothiya) [inline scripts/jenkins/code-coverage.groovy:43]: run st test also

**Review by shailesh-vaidya — APPROVED** [review]
Looks good to me. 
Just one suggestion. Instead of making multiple `sshpass` calls through the Jenkins pipeline, we may create a script like run-code-coverage.sh. Then just copy that script to a remote location and execute it there. This will also remove dependency of Jenkins to run code-coverage.

File: scripts/jenkins/code-coverage.groovy (hunk 2 above)
Comment (rkothiya) [reply scripts/jenkins/code-coverage.groovy:43]: done

File: scripts/jenkins/code-coverage.groovy [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        NODE_PASS = "${NODE_PASS.isEmpty() ? NODE_DEFAULT_SSH_CRED_PSW : NODE_PASS}"
+    }
+
+    stages {
+        stage('Checkout') {
+            steps {
+                checkout([$class: 'GitSCM', branches: [[name: "main"]], doGenerateSubmoduleConfigurations: false, extensions: [[$class: 'PathRestriction', excludedRegions: '', includedRegions: 'scripts/third-party-rpm/.*']], submoduleCfg: [], userRemoteConfigs: [[credentialsId: 'cortx-admin-github', url: "https://github.com/Seagate/cortx-motr"]]])
+            }
+        }
+        stage ('Build rpm packages') {
```
Comment (madhavemuri) [inline scripts/jenkins/code-coverage.groovy:27]: Change the description of the stage here.

**Pull request merged**